### ISSUE TITLE
tcpi_snd_cwnd doesn't seem right on FreeBSD 10.3

### ISSUE THREAD

**Issue by rjmcmahon** [esnet/iperf#465]
I'm looking for FreeBSD expertise w/respect to the net/inetnet/tcp.h The tcpi_snd_cwnd doesn't seem to be right. Iperf3 compiled on FreeBSD 10.3 is reporting an unrealistic congestion window

rjmcmahon@rjm-freebsd ~/Code/iperf-master $ src/iperf3 -c 192.168.100.10 -i 1 -w 1M
Connecting to host 192.168.100.10, port 5201
[ 4] local 192.168.100.22 port 55303 connected to 192.168.100.10 port 5201
[ ID] Interval Transfer Bandwidth Retr Cwnd
[ 4] 0.00-1.00 sec 113 MBytes 949 Mbits/sec 0 294 MBytes
[ 4] 1.00-2.00 sec 112 MBytes 941 Mbits/sec 0 294 MBytes
[ 4] 2.00-3.00 sec 112 MBytes 942 Mbits/sec 0 294 MBytes
[ 4] 3.00-4.00 sec 112 MBytes 941 Mbits/sec 0 294 MBytes
[ 4] 4.00-5.00 sec 112 MBytes 941 Mbits/sec 0 294 MBytes
[ 4] 5.00-6.00 sec 112 MBytes 942 Mbits/sec 0 294 MBytes
[ 4] 6.00-7.00 sec 112 MBytes 941 Mbits/sec 0 294 MBytes
[ 4] 7.00-8.00 sec 112 MBytes 942 Mbits/sec 0 294 MBytes
[ 4] 8.00-9.00 sec 112 MBytes 941 Mbits/sec 0 294 MBytes
[ 4] 9.00-10.00 sec 112 MBytes 941 Mbits/sec 0 294 MBytes

---

[ ID] Interval Transfer Bandwidth Retr
[ 4] 0.00-10.00 sec 1.10 GBytes 942 Mbits/sec 0 sender
[ 4] 0.00-10.00 sec 1.10 GBytes 942 Mbits/sec receiver

iperf Done.

Bob

**Comment by bmah888** [comment]
I've reproduced this problem on FreeBSD 11.0/amd64.  What I think is happening here is that on Linux, `tcp_cwnd` is apparently measured in segments, so we need to multiply that value by `tcp_mss` to get a value in octets.  That's what the function `get_snd_cwnd()` does (modulo the use of -1 as an indication that the platform doesn't support getting this information).

But on FreeBSD, `snd_cwnd` is already in octets.  I can see that by looking (for example) at `tcp_input.c` and just reading the code.  So we're basically seeing integer overflows when we multiply by `snd_mss`.  This (I believe) accounts for the behavior seen in this bug and #475 (which I am going to tentatively and arbitrarily mark as a duplicate of this issue).

I think if I am correct, the fix is pretty straightforward.  That's the good news.

The bad news is that I just FUBARed the FreeBSD VM on my laptop that I was using for my iperf3 development, and I need to find a different machine to use, possibly one of my personal machines.  Stay tuned.

**Comment by bmah888** [comment]
Needs some testing and a merge before we can close this.

**Comment by bmah888** [comment]
Played around with this a bit on a couple of FreeBSD 11 VMs and it seems to be doing the right thing now.  Merging the change to 3.1-STABLE  and closing.

**Comment by rjmcmahon** [comment]
Thanks for doing this.  I'll give it a try on my FreeBSD machine as well

Bob
